File: Graphics/OpenStack.Graphics.OpenGL5/Gl_Shader.cs
//#define DEBUG_SHADERS
using OpenStack.Graphics.Algorithms;
using OpenTK.Graphics.OpenGL;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Text.RegularExpressions;

namespace OpenStack.Graphics.OpenGL
{
    /// <summary>
    /// ShaderLoader
    /// </summary>
    public abstract class ShaderLoader
    {
        const int ShaderSeed = 0x13141516;
        Dictionary<uint, Shader> CachedShaders = [];
        Dictionary<string, List<string>> ShaderDefines = [];

        uint CalculateShaderCacheHash(string name, IDictionary<string, bool> args)
        {
            var b = new StringBuilder(); b.AppendLine(name);
            var parameters = ShaderDefines[name].Intersect(args.Keys);
            foreach (var key in parameters)
            {
                b.AppendLine(key);
                b.AppendLine(args[key] ? "t" : "f");
            }
            return MurmurHash2.Hash(b.ToString(), ShaderSeed);
        }

        protected abstract string GetShaderFileByName(string name);

        protected abstract string GetShaderSource(string name);

        public Shader CreateShader(object path, IDictionary<string, bool> args)
        {
            var name = (string)path;
            var fileName = GetShaderFileByName(name);

            // cache
            if (ShaderDefines.ContainsKey(fileName))
            {
                var shaderCacheHash = CalculateShaderCacheHash(fileName, args);
                if (CachedShaders.TryGetValue(shaderCacheHash, out var c)) return c;
            }

            // build
            var defines = new List<string>();

            // vertex shader
            var vertexShader = GL.CreateShader(ShaderType.VertexShader);
            {
                var shaderSource = GetShaderSource($"{fileName}.vert");
                GL.ShaderSource(vertexShader, PreprocessVertexShader(shaderSource, args));
                // find defines supported from source
                defines.AddRange(FindDefines(shaderSource));
            }
            GL.CompileShader(vertexShader);
            var shaderStatus = GL.GetShaderi(vertexShader, ShaderParameterName.CompileStatus);
            if (shaderStatus != 1)
            {
                GL.GetShaderInfoLog(vertexShader, out var vsInfo);
                throw new Exception($"Error setting up Vertex Shader \"{name}\": {vsInfo}");
            }

            // fragment shader
            var fragmentShader = GL.CreateShader(ShaderType.FragmentShader);
            {
                var shaderSource = GetShaderSource($"{fileName}.frag");
                GL.ShaderSource(fragmentShader, UpdateDefines(shaderSource, args));
                // find render modes supported from source, take union to avoid duplicates
                defines = defines.Union(FindDefines(shaderSource)).ToList();
            }
            GL.CompileShader(fragmentShader);
            shaderStatus = GL.GetShaderi(fragmentShader, ShaderParameterName.CompileStatus);
            if (shaderStatus != 1)
            {
                GL.GetShaderInfoLog(fragmentShader, out var fsInfo);
                throw new Exception($"Error setting up Fragment Shader \"{name}\": {fsInfo}");
            }

            // render modes
            const string RenderMode = "renderMode_";
            var renderModes = defines.Where(k => k.StartsWith(RenderMode)).Select(k => k[RenderMode.Length..]).ToList();

            // build shader
            var shader = new Shader(GL.GetUniformLocation, GL.GetAttribLocation)
            {
                Name = name,
                Parameters = args,
                Program = GL.CreateProgram(),
                RenderModes = renderModes,
            };
            GL.AttachShader(shader.Program, vertexShader);
            GL.AttachShader(shader.Program, fragmentShader);
            GL.LinkProgram(shader.Program);
            GL.ValidateProgram(shader.Program);
            var linkStatus = GL.GetProgrami(shader.Program, ProgramProperty.LinkStatus);
            if (linkStatus != 1)
            {
                GL.GetProgramInfoLog(shader.Program, out var linkInfo);
                throw new Exception($"Error linking shaders: {linkInfo} (link status = {linkStatus})");
            }
            GL.DetachShader(shader.Program, vertexShader);
            GL.DeleteShader(vertexShader);
            GL.DetachShader(shader.Program, fragmentShader);
            GL.DeleteShader(fragmentShader);

#if !DEBUG_SHADERS || !DEBUG
            // cache shader
            ShaderDefines[fileName] = defines;
            var newShaderCacheHash = CalculateShaderCacheHash(fileName, args);
            CachedShaders[newShaderCacheHash] = shader;
            Console.WriteLine($"Shader {newShaderCacheHash} ({name}) ({string.Join(", ", args.Keys)}) compiled and linked succesfully");
#endif
            return shader;
        }

        public Shader CreatePlaneShader(object path, IDictionary<string, bool> args)
        {
            var name = (string)path;
            var shaderFileName = GetShaderFileByName(name);

            // vertex shader
            var vertexShader = GL.CreateShader(ShaderType.VertexShader);
            {
                var shaderSource = GetShaderSource($"plane.vert");
                GL.ShaderSource(vertexShader, shaderSource);
            }
            GL.CompileShader(vertexShader);
            var shaderStatus = GL.GetShaderi(vertexShader, ShaderParameterName.CompileStatus);
            if (shaderStatus != 1)
            {
                GL.GetShaderInfoLog(vertexShader, out var vsInfo);
                throw new Exception($"Error setting up Vertex Shader \"{name}\": {vsInfo}");
            }

            // fragment shader
            var fragmentShader = GL.CreateShader(ShaderType.FragmentShader);
            {
                var shaderSource = GetShaderSource($"{shaderFileName}.frag");
                GL.ShaderSource(fragmentShader, UpdateDefines(shaderSource, args));
            }
            GL.CompileShader(fragmentShader);
            shaderStatus = GL.GetShaderi(fragmentShader, ShaderParameterName.CompileStatus);
            if (shaderStatus != 1)
            {
                GL.GetShaderInfoLog(fragmentShader, out var fsInfo);
                throw new Exception($"Error setting up Fragment Shader \"{name}\": {fsInfo}");
            }

            // build shader
            var shader = new Shader(GL.GetUniformLocation, GL.GetAttribLocation)
            {
                Name = name,
                Program = GL.CreateProgram(),
            };
            GL.AttachShader(shader.Program, vertexShader);
            GL.AttachShader(shader.Program, fragmentShader);
            GL.LinkProgram(shader.Program);
            GL.ValidateProgram(shader.Program);
            var linkStatus = GL.GetProgrami(shader.Program, ProgramProperty.LinkStatus);
            if (linkStatus != 1)
            {
                GL.GetProgramInfoLog(shader.Program, out var linkInfo);
                throw new Exception($"Error linking shaders: {linkInfo} (link status = {linkStatus}");
            }
            GL.DetachShader(shader.Program, vertexShader);
            GL.DeleteShader(vertexShader);
            GL.DetachShader(shader.Program, fragmentShader);
            GL.DeleteShader(fragmentShader);
            return shader;
        }

        // Preprocess a vertex shader's source to include the #version plus #defines for parameters
        string PreprocessVertexShader(string source, IDictionary<string, bool> args)
            => ResolveIncludes(UpdateDefines(source, args));

        // Update default defines with possible overrides from the model
        static string UpdateDefines(string source, IDictionary<string, bool> args)
        {
            // find all #define param_(paramName) (paramValue) using regex
            var defines = Regex.Matches(source, @"#define param_(\S*?) (\S*?)\s*?\n");
            foreach (Match define in defines)
                // check if this parameter is in the arguments
                if (args.TryGetValue(define.Groups[1].Value, out var value))
                {
                    // overwrite default value
                    var index = define.Groups[2].Index;
                    var length = define.Groups[2].Length;
                    source = source.Remove(index, Math.Min(length, source.Length - index)).Insert(index, value ? "1" : "0");
                }
            return source;
        }

        // Remove any #includes from the shader and replace with the included code
        string ResolveIncludes(string source)
        {
            var includes = Regex.Matches(source, @"#include ""([^""]*?)"";?\s*\n");
            foreach (Match define in includes)
            {
                // read included code
                var includedCode = GetShaderSource(define.Groups[1].Value);
                // recursively resolve includes in the included code. (Watch out for cyclic dependencies!)
                includedCode = ResolveIncludes(includedCode);
                if (!includedCode.EndsWith("\n")) includedCode += "\n";
                // replace the include with the code
                source = source.Replace(define.Value, includedCode);
            }
            return source;
        }

        static List<string> FindDefines(string source)
        {
            var defines = Regex.Matches(source, @"#define param_(\S+)");
            return defines.Cast<Match>().Select(_ => _.Groups[1].Value).ToList();
        }
    }

    /// <summary>
    /// ShaderDebugLoader
    /// </summary>
    public class ShaderDebugLoader : ShaderLoader
    {
        const string ShaderDirectory = "OpenStack.Graphics.OpenGL5.Shaders";

        // Map shader names to shader files
        protected override string GetShaderFileByName(string name)
        {
            switch (name)
            {
                case "plane": return "plane";
                case "vrf.error": return "error";
                case "vrf.grid": return "debug_grid";
                case "vrf.picking": return "picking";
                case "vrf.particle.sprite": return "particle_sprite";
                case "vrf.particle.trail": return "particle_trail";
                case "tools_sprite.vfx": return "sprite";
                case "vr_unlit.vfx": return "vr_unlit";
                case "vr_black_unlit.vfx": return "vr_black_unlit";
                case "water_dota.vfx": return "water";
                case "hero.vfx":
                case "hero_underlords.vfx": return "dota_hero";
                case "multiblend.vfx": return "multiblend";
                default:
                    if (name.StartsWith("vr_")) return "vr_standard";
                    // Console.WriteLine($"Unknown shader {name}, defaulting to simple.");
                    return "simple";
            }
        }

        protected override string GetShaderSource(string name)
        {
#if DEBUG_SHADERS && DEBUG
            using (var stream = File.Open(GetShaderDiskPath(name), FileMode.Open))
#else
            var assembly = Assembly.GetExecutingAssembly();
            using (var stream = assembly.GetManifestResourceStream($"{ShaderDirectory}.{name}"))
#endif
            using (var reader = new StreamReader(stream)) return reader.ReadToEnd();
        }

#if DEBUG_SHADERS && DEBUG
        // Reload shaders at runtime
        static string GetShaderDiskPath(string name) => Path.Combine(Path.GetDirectoryName(Process.GetCurrentProcess().MainModule?.FileName), "../../../../", ShaderDirectory.Replace(".", "/"), name);
#endif
    }
}

File: Python/openstk/gfx/gl_shader.py
import re, math, numpy as np
from importlib import resources
from OpenGL.GL import *
from openstk.gfx_render import Shader

# ShaderSeed = 0x13141516
RenderMode = 'renderMode_'; RenderModeLength = len(RenderMode)

# ShaderLoader
class ShaderLoader:
    _cachedShaders: dict[int, Shader] = {}
    _shaderDefines: dict[str, list[str]] = {}
    
    def _calculateShaderCacheHash(self, name: str, args: dict[str, bool]) -> int:
        b = [name]
        parameters = self._shaderDefines[name].intersect(args.keys)
        for key in parameters:
            b.append(key)
            b.append('t' if args[key] else 'f')
        return hash('\n'.join(b))

    def getShaderFileByName(self, name: str) -> str: pass

    def getShaderSource(self, name: str) -> str: pass

    def createShader(self, path: object, args: dict[str, bool]) -> Shader:
        name = str(path)
        fileName = self.getShaderFileByName(name)
        
        # cache
        if fileName in self._shaderDefines:
            shaderCacheHash = self._calculateShaderCacheHash(fileName, args)
            if shaderCacheHash in self._cachedShaders: return self._cachedShaders[shaderCacheHash]

        # build
        defines = []

        # vertex shader
        vertexShader = glCreateShader(GL_VERTEX_SHADER)
        if True:
            shaderSource = self.getShaderSource(f'{fileName}.vert')
            glShaderSource(vertexShader, self.preprocessVertexShader(shaderSource, args))
            # find defines supported from source
            defines += self.findDefines(shaderSource)
        glCompileShader(vertexShader)
        shaderStatus = glGetShaderiv(vertexShader, GL_COMPILE_STATUS)
        if shaderStatus != 1:
            vsInfo = glGetShaderInfoLog(vertexShader)
            raise Exception(f'Error setting up Vertex Shader "{name}": {vsInfo}')

        # fragment shader
        fragmentShader = glCreateShader(GL_FRAGMENT_SHADER)
        if True:
            shaderSource = self.getShaderSource(f'{fileName}.frag')
            glShaderSource(fragmentShader, self.updateDefines(shaderSource, args))
            # find render modes supported from source, take union to avoid duplicates
            defines += self.findDefines(shaderSource)
        glCompileShader(fragmentShader)
        shaderStatus = glGetShaderiv(fragmentShader, GL_COMPILE_STATUS)
        if shaderStatus != 1:
            fsInfo = glGetShaderInfoLog(fragmentShader)
            raise Exception(f'Error setting up Fragment Shader "{name}": {fsInfo}')

        # render modes
        renderModes = [k[RenderModeLength] for k in defines if k.startswith(RenderMode)]

        # build shader
        shader = Shader(glGetUniformLocation, glGetAttribLocation,
            name = name,
            parameters = args,
            program = glCreateProgram(),
            renderModes = renderModes)
        glAttachShader(shader.program, vertexShader)
        glAttachShader(shader.program, fragmentShader)
        glLinkProgram(shader.program)
        glValidateProgram(shader.program)
        linkStatus = glGetProgramiv(shader.program, GL_LINK_STATUS)
        if linkStatus != 1:
            linkInfo = glGetProgramInfoLog(shader.program)
            raise Exception(f'Error linking shaders: {linkInfo} (link status = {linkStatus})')
        glDetachShader(shader.program, vertexShader)
        glDeleteShader(vertexShader)
        glDetachShader(shader.program, fragmentShader)
        glDeleteShader(fragmentShader)

        # cache shader
        if True:
            self._shaderDefines[fileName] = defines
            newShaderCacheHash = self._calculateShaderCacheHash(fileName, args)
            self._cachedShaders[newShaderCacheHash] = shader
            print(f'Shader {newShaderCacheHash} ({name}) ({', '.join(args.Keys)}) compiled and linked succesfully')
        return shader

    def createPlaneShader(self, path: object, args: dict[str, bool]) -> Shader:
        name = str(path)
        fileName = self.getShaderFileByName(name)

        # vertex shader
        vertexShader = glCreateShader(GL_VERTEX_SHADER)
        if True:
            shaderSource = self.getShaderSource('plane.vert')
            glShaderSource(vertexShader, shaderSource)
        glCompileShader(vertexShader)
        shaderStatus = glGetShaderiv(vertexShader, GL_COMPILE_STATUS)
        if shaderStatus != 1:
            vsInfo = glGetShaderInfoLog(vertexShader)
            raise Exception(f'Error setting up Vertex Shader "{name}": {vsInfo}')

        # fragment shader
        fragmentShader = glCreateShader(GL_FRAGMENT_SHADER)
        if True:
            shaderSource = self.getShaderSource(f'{fileName}.frag')
            glShaderSource(fragmentShader, self.updateDefines(shaderSource, args))
        glCompileShader(fragmentShader)
        shaderStatus = glGetShaderiv(fragmentShader, GL_COMPILE_STATUS)
        if shaderStatus != 1:
            fsInfo = glGetShaderInfoLog(fragmentShader)
            raise Exception(f'Error setting up Fragment Shader "{name}": {fsInfo}')

        # build shader
        shader = Shader(glGetUniformLocation, glGetAttribLocation,
            name = name,
            program = glCreateProgram())
        glAttachShader(shader.program, vertexShader)
        glAttachShader(shader.program, fragmentShader)
        glLinkProgram(shader.program)
        glValidateProgram(shader.program)
        linkStatus = glGetProgramiv(shader.program, GL_LINK_STATUS)
        if linkStatus != 1:
            linkInfo = glGetProgramInfoLog(shader.program)
            raise Exception(f'Error linking shaders: {linkInfo} (link status = {linkStatus})')
        glDetachShader(shader.program, vertexShader)
        glDeleteShader(vertexShader)
        glDetachShader(shader.program, fragmentShader)
        glDeleteShader(fragmentShader)
        return shader

    # Preprocess a vertex shader's source to include the #version plus #defines for parameters
    def preprocessVertexShader(self, source: str, args: dict[str, bool]) -> str:
        return self.resolveIncludes(self.updateDefines(source, args))

    # Update default defines with possible overrides from the model
    @staticmethod
    def updateDefines(source: str, args: dict[str, bool]) -> str:
        # find all #define param_(paramName) (paramValue) using regex
        defines = re.compile('#define param_(\\S*?) (\\S*?)\\s*?\\n').finditer(source)
        for define in defines:
            # check if this parameter is in the arguments
            if (key := define[1]) in args:
                # overwrite default value
                start, end = define.span(2)
                source = source[:start] + ('1' if args[key] else '0') + source[end:]
        return source

    # Remove any #includes from the shader and replace with the included code
    def resolveIncludes(self, source: str) -> str:
        includes = re.compile('#include "([^"]*?)";?\\s*\\n').finditer(source)
        for define in includes:
            # read included code
            includedCode = self.getShaderSource(define[1])
            # recursively resolve includes in the included code. (Watch out for cyclic dependencies!)
            includedCode = self.resolveIncludes(includedCode)
            if not includedCode.endswith('\n'): includedCode += '\n'
            # replace the include with the code
            start, end = define.span(0)
            source = source.replace(source[start:end], includedCode)
        return source

    @staticmethod
    def findDefines(source: str) -> list[str]:
        defines = re.compile('#define param_(\\S+)').finditer(source)
        return [x[1] for x in defines]

# ShaderDebugLoader
class ShaderDebugLoader(ShaderLoader):
    def getShaderFileByName(self, name: str) -> str:
        match name:
            case 'plane': return 'plane'
            case 'vrf.error': return 'error'
            case 'vrf.grid': return 'debug_grid'
            case 'vrf.picking': return 'picking'
            case 'vrf.particle.sprite': return 'particle_sprite'
            case 'vrf.particle.trail': return 'particle_trail'
            case 'tools_sprite.vfx': return 'sprite'
            case 'vr_unlit.vfx': return 'vr_unlit'
            case 'vr_black_unlit.vfx': return 'vr_black_unlit'
            case 'water_dota.vfx': return 'water'
            case 'hero.vfx', 'hero_underlords.vfx': return 'dota_hero'
            case 'multiblend.vfx': return 'multiblend'
            case _:
                if name.startsWith('vr_'): return 'vr_standard'
                return 'simple'

    def getShaderSource(self, name: str) -> str:
        return resources.files().joinpath('shaders', name).read_text(encoding='utf-8')

# print(ShaderDebugLoader().loadShader('water_dota.vfx', {'fulltangent': 1}))